### ISSUE TITLE
Feature Request: being able to dynamically choose database or import/export browser local data

### ISSUE THREAD

**Issue by khaledvic** [vivekratnavel/omniboard#69]
Since my mongo database become bigger, I'm splitting each group of related experiments to their own database.  I have to choose different port for each database because of the local data Omniboard stores for each database in the browser. However, as the number of databases become larger it becomes harder to keep track of ports (which for which database).

I don't know what are the best practice to deal with this but I can think of either:

- Making Omniboard somehow similar to Jupyter, so you see all the databases (like you see Jupyter  notebooks in a folder) and then by selecting a database an instance of Omniboard is run on that database with its own configuration (browser local data)

- Being able to export import the local-data, so I can reuse the same port for multiple datasets.

- Storing this local data in the mongo database with an option to restore/save it to the local browser data.

**Comment by vivekratnavel** [comment]
@khaledvic 
`Saving views in mongo db with an option to switch between views` looks good to me.

Saving a view in Omniboard will save the following:
- Column order
- Show/hide columns
- Column resizing
- Column sort order
- Status filters and advanced filters

Is there anything more you would like to add?

**Comment by khaledvic** [comment]
That looks good. I think also it may be useful to store custom and metrics columns

**Comment by vivekratnavel** [comment]
Custom config columns and metric columns are already being saved in Mongodb. They won't be part of a view.

**Comment by albertocarot1** [comment]
That would be a very useful feature indeed! 
especially when omniboard is served.

**Comment by nuth** [comment]
I've created a patch in my own fork at https://github.com/nuth/omniboard/pull/1 which I think could do what you want, by connecting to multiple databases, and expose each on their own path. It's a bit crudely done, but seems to work well.

It requires a new config file that looks like this:
`{
  "sacred": {
    "mongoUri": "mongodb://mongo/sacred",
    "path":"/sacred"
  },
  "sacred2": {
    "mongoUri": "mongodb://mongo/sacred2",
    "path":"/sacred2"
  }
}`

There's a few things I'd like to fix before I do a pull request, but if you'd like to try it out you can run it like this:
`docker run -v $PWD/config.json:/usr/omniboard/config.json  --net=sacred-network --rm -p 9000:9000 nuthman/omniboard`

@vivekratnavel something you'd be interested in?

**Comment by vivekratnavel** [comment]
@nuth I like this approach of having a config file. Your patch might need some more polishing and testing before we can merge. I am currently working on v2 #107 which involves a lot of code changes in both the server side and client side and I don't have time to review or test this approach. I would definitely be interested to take a look once v2 is released, which I am hoping to do in a couple weeks.

@u39kun What do you think about this approach to support multiple databases?

**Comment by vivekratnavel** [comment]
Also, I forgot to mention that the way state is stored in localstorage might cause an issue here. We will need to modify the localstorage keys to include paths like `/sacred`, `/sacred2`.

**Comment by vivekratnavel** [comment]
Omniboard v2.0.0-beta is out. You can try it out with either npm or docker - https://www.npmjs.com/package/omniboard/v/2.0.0-beta.2 
Docker image - vivekratnavel/omniboard:beta
Source code is in branch 'beta' - https://github.com/vivekratnavel/omniboard/tree/beta

Please try it out and give feedback.

**Comment by vivekratnavel** [comment]
@nuth Can you test the config changes against the branch 'beta' and create a PR against 'beta'?

I can now test the config changes and we can release it as part of v2.0.0

Thanks for your time and effort!

**Comment by nuth** [comment]
@vivekratnavel cool! I'll look at it ASAP, so in a week or two, since I'm deep into something else atm.

**Comment by lotsoffood** [comment]
Any movement on this? I'm stoked to be able to have multiple databases selectable from within Omniboard.

**Comment by nuth** [comment]
Thanks for the reminder! I'll have some time right over new years to merge
with current and finish up, and make a pull request and then we'll see how
it goes :-)

søn. 29. des. 2019, 16:14 skrev lotsoffood <notifications@github.com>:

**Comment by lotsoffood** [comment]
@nuth Let me know if you need some help with the testing or point me to the branch it's on.

**Comment by nuth** [comment]
> @nuth Let me know if you need some help with the testing or point me to the branch it's on.

Yes, that'd be very nice, since I seem to always end up with my hands full lately. The code is here: https://github.com/nuth/omniboard/pull/1

What's missing is:
* Merge in master (v2)
* Dynamically set location of config file through environment variable (12 factor)
* Any improvements to the config file format
* Sort out any frontend quirks, like local storage per context path, etc.

If you have time, I'd be very happy to hand it over, and I'm of course available to answer any questions :)

**Comment by rohithkumar31** [comment]
Any update on this feature?

**Comment by gkanwar** [comment]
I also have a similar use case with multiple sacred databases. Is this feature being actively worked on? What remaining tasks need to be handled before merging? I could spend a few hours trying to move this along.

**Comment by nuth** [comment]
> I also have a similar use case with multiple sacred databases. Is this feature being actively worked on? What remaining tasks need to be handled before merging? I could spend a few hours trying to move this along.

As far as I know, same status as my last post https://github.com/vivekratnavel/omniboard/issues/69#issuecomment-571746879

I don't think I'll have time to work on the project anymore, since I no longer work at the same project.

**Comment by gkanwar** [comment]
Okay, thanks @nuth, working on picking up from your MP-6140 branch now. Is there a reason for an environment variable for the config file instead of a command line argument?

@vivekratnavel are there key improvements you would like to see before this can be merged? I would like to make this available for my project as soon as possible :grin:

**Comment by nuth** [comment]
The solution seems to follow 12 factors (https://12factor.net/) more or
less, so it should be configurable through an environment variable, which
is a good idea anyways when running things in docker.

**Comment by gkanwar** [comment]
Seems like the 12 factor approach would be not to have a config file at all, but set all this information directly in environment variables?

> The twelve-factor app stores config in environment variables (often shortened to env vars or env). Env vars are easy to change between deploys without changing any code; unlike config files, there is little chance of them being checked into the code repo accidentally; and unlike custom config files, or other config mechanisms such as Java System Properties, they are a language- and OS-agnostic standard.

**Comment by nuth** [comment]
Indeed, it's not perfect, but in my mind hard coding the config file would
be less in line with 12 factor than a config file controlled by an
environment variable... of course you could store the contents of the
config file in an environment variable, but the readability of that is
terrible. 12 factor is after all just a guideline for creating readable,
flexible code/apps, and I'd say in this instance, not following it to the
dot makes it more readable.

As for why not creating this as purely environmental variables, I feel the
structure doesn't really fit all that well in to a strict key value format.

Of course, now you're in charge, so feel free to finish it however you like
:)

**Comment by gkanwar** [comment]
Haha, I'm happy to stick to this, just wanted to understand the motivations. Getting a bit of headache trying to merge app.js right now... but getting close!

**Comment by gkanwar** [comment]
Was this code passing the web tests on your branch? I'm finding that e.g. in runsTable.test.js the expects around mockAxios want a leading '/'. Does this make sense?

**Comment by gkanwar** [comment]
Merge complete, and opened PR #200 so we can iterate on remaining issues. It seems like environment variable support was already there, taking config file path from OMNIBOARD_CONFIG. I've tested locally versus multiple sacred DBs with different datasets and basic operations like displaying list of runs and opening the drill down view seem to work fine.

**Comment by nuth** [comment]
> Was this code passing the web tests on your branch? I'm finding that e.g. in runsTable.test.js the expects around mockAxios want a leading '/' like '/api/...'. Does this make sense?

Ahh, yes... I might have skipped the tests. I think they all should work, if you just account for the updated context path stuff.

**Comment by gkanwar** [comment]
Seems like some mockAxios calls want "/api/..." and some want "api/..." and I don't understand when or why. Could you explain what is happening in `Backend/backend.js`? How does it eventually learn about what the current path is?
